staging/src/k8s.io/apiserver/pkg/cel/library/ip.go

```diff
@@ -0,0 +1,307 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"fmt"
+	"net/netip"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// IPs provides a CEL function library extension of IP address parsing functions.
+//
+// ip
+//
+// Converts a string to an IP address or results in an error if the string is not a valid IP address.
+// The IP address must be an IPv4 or IPv6 address.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4) are not allowed.
+// IP addresses with zones (e.g. fe80::1%eth0) are not allowed.
+// Leading zeros in IPv4 address octets are not allowed.
+//
+//	ip(<string>) <IPAddr>
+//
+// Examples:
+//
+//	ip('127.0.0.1') // returns an IPv4 address
+//	ip('::1') // returns an IPv6 address
+//	ip('127.0.0.256') // error
+//	ip(':::1') // error
+//
+// isIP
+//
+// Returns true if a string is a valid IP address.
+// The IP address must be an IPv4 or IPv6 address.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4) are not allowed.
+// IP addresses with zones (e.g. fe80::1%eth0) are not allowed.
+// Leading zeros in IPv4 address octets are not allowed.
+//
+//	isIP(<string>) <bool>
+//
+// Examples:
+//
+//	isIP('127.0.0.1') // returns true
+//	isIP('::1') // returns true
+//	isIP('127.0.0.256') // returns false
+//	isIP(':::1') // returns false
+//
+// family / isCanonical / isInCIDR / isUnspecified / isLoopback / isLinkLocalMulticast / isLinkLocalUnicast
+//
+// Returns true if a string is a valid IPv4 address, IPv6 address, or is in its canonical format.
+//
+// - family: returns the IP addresses' family (IPv4 or IPv6) as an integer, either '4' or '6'.
+//
+// - isCanonical: returns true when the IP Address is in its canonical format. There is exactly
+// one canonical form for every IP address, so fields containing IPs in canonical
+// form can just be treated as strings when checking for equality or uniqueness.
+//
+// - isInCIDR: returns true if the IP address is contained in the CIDR.
+// The CIDR must be an IPv4 or IPv6 subnet address with a mask, parsed with cidr.
+//
+// - isUnspecified: returns true if the IP address is the unspecified address.
+// Either the IPv4 address "0.0.0.0" or the IPv6 address "::".
+//
+// - isLoopback: returns true if the IP address is the loopback address.
+// Either an IPv4 address with a value of 127.x.x.x or an IPv6 address with a value of ::1.
+//
+// - isLinkLocalMulticast: returns true if the IP address is a link-local multicast address.
```

Comment:
> "special" is not a standard defintion, so it's not a great candidate for this CEL library (I assume). So where SHOULD it go? "Special" is a kube-centric, definition, though.

This PR is kube-centric. (It's not changing "core CEL" or whatever, it's defining kube-specific extensions.)

But also, `ValidateNonSpecialIP` is nearly identical to `net.IP.IsGlobalUnicast` / `netip.Addr.IsGlobalUnicast`. (`ValidateNonSpecialIP` allows some broadcast and multicast IPs that `IsGlobalUnicast` doesn't, but that's probably a bug.) So we could just call it `isGlobalUnicast`, using the same checks as the golang stdlib, and that would make it less ambiguous what it's supposed to mean.

Currently we only use `ValidateNonSpecialIP` for service externalIPs, endpoint IPs, and endpointslice IPs. It probably ought to be used more than that...

Reply:
IsGlobalUnicast seems OK, but we need to make sure it's not a breaking change :)

Reply:
> Or (just thinking ouot loud) is there a way to define "local" functions that get loaded for some evaluations? e.g. in my types.go, somewhere near the top:

To my knowledge, this doesn't exist (yet?). I've not seen any talk of it in the KEP either.

I can add an `IsGlobablUnicast`. I don't think it would be breaking since nothing is using this now, all of this validation is net new. The breaking change would be changing something that uses `ValidateNonSpecialIP` to this without accounting for the differences between them. Are we expecting anyone to do that or is the helper only used on built in types?

Reply:
Tim is worrying about the bright shiny future where we drop all the golang validation and replace it all with CEL. If that happens, we will probably need to add an `isLegacyNonSpecialIP()` method specifically for the objects that are currently using `ValidateNonSpecialIP`, because there's no other way to express *exactly* what that function currently checks unless we export every single existing IP-testing function (`IsLinkLocalMulticast` etc).